### PR TITLE
[ThinLTO] Allow importing based on a workload definition

### PR DESCRIPTION
**Pull request by mtrofin**
An example of a "workload definition" would be "the transitive closure of functions actually called to satisfy a RPC request", i.e. a (typically significantly) smaller subset of the transitive closure (static + possible indirect call targets) of callees. This means this workload definition is a type of flat dynamic profile.

Producing one is not in scope - it can be produced offline from traces, or from sample-based profiles, etc.

This patch adds awareness to ThinLTO of such a concept. A workload is defined as a root and a list of functions. All function references are by-name (more readable than GUIDs). In the case of aliases, the expectation is the list contains all the alternative names.

The workload definitions are presented to the linker as a json file, containing a dictionary. The keys are the roots, the values are the list of functions.

The import list for a module defining a root will be the functions listed for it in the profile.

Using names this way assumes unique names for internal functions, i.e. clang's `-funique-internal-linkage-names`.

Note that the behavior affects the entire module where a root is defined (i.e. different workloads best be defined in different modules), and does not affect modules that don't define roots.

### CODE REVIEW

File: llvm/test/ThinLTO/X86/workload.ll [hunk 1]
```diff
@@ -0,0 +1,67 @@
+; RUN: mkdir -p %t_baseline
+; RUN: opt -module-summary %S/Inputs/workload1.ll -o %t_baseline/workload1.bc
```
Comment (teresajohnson) [inline llvm/test/ThinLTO/X86/workload.ll:2]: Recommend including all the inputs here and use split-file (https://llvm.org/docs/TestingGuide.html#extra-files). That would make it easier to follow what is going on, especially since there is no code in this file currently.

File: llvm/test/ThinLTO/X86/workload.ll [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+; RUN: opt -module-summary %S/Inputs/workload2.ll -o %t_baseline/workload2.bc
+; RUN: opt -module-summary %S/Inputs/workload3.ll -o %t_baseline/workload3.bc
+;
+; Normal run. The first module shouldn't get m2_f1
```
Comment (teresajohnson) [inline llvm/test/ThinLTO/X86/workload.ll:6]: because it isn't referenced from there (?)

File: llvm/test/ThinLTO/X86/workload.ll (hunk 1 above)
Comment (teresajohnson) [reply llvm/test/ThinLTO/X86/workload.ll:2]: Also, the test is a little confusing because some of the specified imports (e.g. m1_f1 lists m2_f1 and vice versa) are not actually (transitively) referenced from the importing module. I guess this is just for testing purposes, but it would be good to have a test that shows/tests the usage for a case where the imports would normally be expected.

File: llvm/test/ThinLTO/X86/workload.ll [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+; while keeping its variant of noninterposable_f
+;
+; FIRST-LABEL: @m1_f1
+; FIRST-LABEL: @m1_f2
+; FIRST-LABEL: define available_externally void @noninterposable_f
+; FIRST-NEXT: call void @m1_variant
+; FIRST-LABEL: @m2_f1
+; FIRST-LABEL: define available_externally void @interposable_f
+; FIRST-NEXT: call void @m2_variant
+; FIRST-NOT:   @m2_f1_alias
```
Comment (teresajohnson) [inline llvm/test/ThinLTO/X86/workload.ll:63]: Why isn't m2_f1_alias imported, since it was specified in the import list of m1? I guess that is due to the way the importer handles imports (?) - it would be good to have a comment as to why.

File: llvm/test/ThinLTO/X86/Inputs/workload2.ll [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+define dso_local void @m2_f1() {
+  call void @interposable_f()
+  call void @noninterposable_f()
+  ret void
+}
+
+@m2_f1_alias = alias void (...), ptr @m2_f1
+
+define linkonce_odr void @interposable_f() {
```
Comment (teresajohnson) [inline llvm/test/ThinLTO/X86/Inputs/workload2.ll:14]: Is this meant to have interposable linkage?

File: llvm/test/ThinLTO/X86/workload.ll [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+; RUN:  -r %t/workload3.bc,m1_f1 \
+; RUN:  -r %t/workload3.bc,m3_f1,plx
+; RUN: llvm-dis %t/result.o.1.3.import.bc -o - | FileCheck %s --check-prefix=FIRST
+; RUN: llvm-dis %t/result.o.2.3.import.bc -o - | FileCheck %s --check-prefix=SECOND
+; RUN: llvm-dis %t/result.o.3.3.import.bc -o - | FileCheck %s --check-prefix=THIRD
+;
+; The third module is bitwse-identical to the "normal" run, as the 
+; RUN: diff %t_baseline/result.o.3.3.import.bc %t/result.o.3.3.import.bc
+;
+; This time, we expect m1 to have m2_f1 and the m2 variant of interposable_f,
```
Comment (teresajohnson) [inline llvm/test/ThinLTO/X86/workload.ll:53]: It would be good to have all these checks and the comments as to what was imported vs not up earlier right after the workload json file is written. It's a little hard to follow right now because the code is in different files and the workload file is written in a different part of this file.

**Review by teresajohnson — COMMENTED** [review]
Sorry for the delay. Mostly suggestions for improving the clarity of what is happening in the code and test.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp [hunk 6]
```diff
@@ -369,29 +373,238 @@ class GlobalsImporter final {
...
+  static std::unique_ptr<ModuleImportsManager>
+  create(function_ref<bool(GlobalValue::GUID, const GlobalValueSummary *)>
+             IsPrevailing,
+         const ModuleSummaryIndex &Index,
+         DenseMap<StringRef, FunctionImporter::ExportSetTy> *ExportLists =
+             nullptr);
 };
 
+class WorkloadImportsManager : public ModuleImportsManager {
+  // Keep a module name -> defined value infos association. We use it to
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:413]: Correct.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp [hunk 7]
```diff
@@ -369,29 +373,238 @@ class GlobalsImporter final {
...
+        LLVM_DEBUG(dbgs() << "[Workload] Not importing " << VI.name()
+                          << " because can't select Callee. Guid is: "
+                          << Function::getGUID(VI.name())
+                          << ". The reason was: " << getFailureName(LastReason)
+                          << "\n");
+        continue;
+      }
+      const auto *CFS = cast<FunctionSummary>(GVS->getBaseObject());
+      auto ExportingModule = CFS->modulePath();
+      if (ExportingModule == ModName) {
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:481]: It's mostly for debugging / clarity.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp [hunk 8]
```diff
@@ -369,29 +373,238 @@ class GlobalsImporter final {
...
+    LLVM_DEBUG(dbgs() << "[Workload] Done\n");
+  }
+
+  bool shouldImport(const GVSummaryMapTy &DefinedGVSummaries,
+                    Function::GUID Guid, const GlobalValueSummary *Candidate) {
+    auto DefinedSummary = DefinedGVSummaries.find(Guid);
+    if (DefinedSummary == DefinedGVSummaries.end())
+      return true;
+
+    // See shouldImportGlobal for the justificaton of the isInterposableLinkage.
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:510]: adding comment in the overall candidate selection logic.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp [hunk 9]
```diff
@@ -369,29 +373,238 @@ class GlobalsImporter final {
...
+  WorkloadImportsManager(
+      function_ref<bool(GlobalValue::GUID, const GlobalValueSummary *)>
+          IsPrevailing,
+      const ModuleSummaryIndex &Index,
+      DenseMap<StringRef, FunctionImporter::ExportSetTy> *ExportLists)
+      : ModuleImportsManager(IsPrevailing, Index, ExportLists) {
+    StringMap<ValueInfo> CtxGuidToValueInfo;
+    for (auto &I : Index) {
+      ValueInfo VI(Index.haveGVs(), &I);
+      CtxGuidToValueInfo[VI.name()] = VI;
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:534]: Added the comments. For test, @m1_f2 is `internal` and imported by the second module.

[the remark this answers is not in the document]
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:534]: Also reporting (LLVM_DBGS) ambiguous names, if the workflow def references such things.

File: llvm/test/ThinLTO/X86/Inputs/workload2.ll (hunk 4 above)
Comment (mtrofin) [reply llvm/test/ThinLTO/X86/Inputs/workload2.ll:14]: Well, not *this* one. Its similarly named one in workload1.ll does. The test checks that this copy here will replace the one in workload1.ll.

File: llvm/test/ThinLTO/X86/workload.ll (hunk 1 above)
Comment (mtrofin) [reply llvm/test/ThinLTO/X86/workload.ll:2]: Right, my intention was to show that for a few functions, they'll get imported "regardless". m3 pulls from m1 "normally".

File: llvm/test/ThinLTO/X86/workload.ll [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+; RUN:  -r %t/workload2.bc,interposable_f,p \
+; RUN:  -r %t/workload2.bc,noninterposable_f,p \
+; RUN:  -r %t/workload2.bc,m2_variant \
+; RUN:  -r %t/workload3.bc,m1_f1 \
+; RUN:  -r %t/workload3.bc,m3_f1,plx
+; RUN: llvm-dis %t/result.o.1.3.import.bc -o - | FileCheck %s --check-prefix=FIRST
+; RUN: llvm-dis %t/result.o.2.3.import.bc -o - | FileCheck %s --check-prefix=SECOND
+; RUN: llvm-dis %t/result.o.3.3.import.bc -o - | FileCheck %s --check-prefix=THIRD
+;
+; The third module is bitwse-identical to the "normal" run, as the
```
[the remark this answers is not in the document]
Comment (teresajohnson) [reply llvm/test/ThinLTO/X86/workload.ll:50]: This is marked resolved but I still see an incomplete sentence: "The third module is bitwse-identical to the "normal" run, as the"

File: llvm/lib/Transforms/IPO/FunctionImport.cpp [hunk 11]
```diff
@@ -369,29 +387,273 @@ class GlobalsImporter final {
...
+      FunctionImporter::ImportFailureReason LastReason =
+          FunctionImporter::ImportFailureReason::None;
+      for (const auto &Candidate : Candidates) {
+        /// We will prefer importing the prevailing candidate, if not, we'll
+        /// still pick the first available candidate. The reason we want to make
+        /// sure we do import the prevailing candidate is because the goal of
+        /// workload-awareness is to enable optimizations specializing the call
+        /// graph of that workload. Suppose a function is already defined in the
+        /// module, but it's not the prevailing variant. Suppose also we do not
+        /// inline it, but we could specialize it to the workload in other ways.
```
Comment (teresajohnson) [inline llvm/lib/Transforms/IPO/FunctionImport.cpp:467]: In fact we *cannot* inline interposable functions.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp [hunk 12]
```diff
@@ -369,29 +387,273 @@ class GlobalsImporter final {
...
+      auto Candidates =
+          qualifyCalleeCandidates(Index, VI.getSummaryList(), ModName);
+
+      const GlobalValueSummary *GVS = nullptr;
+      FunctionImporter::ImportFailureReason LastReason =
+          FunctionImporter::ImportFailureReason::None;
+      for (const auto &Candidate : Candidates) {
+        /// We will prefer importing the prevailing candidate, if not, we'll
+        /// still pick the first available candidate. The reason we want to make
+        /// sure we do import the prevailing candidate is because the goal of
```
Comment (teresajohnson) [inline llvm/lib/Transforms/IPO/FunctionImport.cpp:463]: Add note about this only being for interposable linkage, and why we don't do the same for non-interposable (e.g. linkonce_odr)

File: llvm/test/ThinLTO/X86/Inputs/workload2.ll (hunk 4 above)
Comment (teresajohnson) [reply llvm/test/ThinLTO/X86/Inputs/workload2.ll:14]: Maybe make it have strong (external) linkage then. I'm not sure in reality that the linker would ever select a linkonce_odr as prevailing over linkonce. Typically it is an interposable (e.g. linkonce or weak) symbol being overridden by a strong symbol.

File: llvm/test/ThinLTO/X86/workload.ll (hunk 10 above)
[the remark this answers is not in the document]
Comment (mtrofin) [reply llvm/test/ThinLTO/X86/workload.ll:50]: VSCode eagerly sends the "resolved conversation" as I make local changes (i.e. before I actually push the changes)

File: llvm/test/ThinLTO/X86/workload.ll (hunk 3 above)
Comment (mtrofin) [reply llvm/test/ThinLTO/X86/workload.ll:63]: It should be actually. (summarizing offline discussion): We need to assume aliases and aliasees are given in the import list because the source of the workload def may be a tool working off sections in the binary, and there's no information as to which name is alias or aliasee - all we know is that multiple symbols map to the same address. That aside, we will then import what's given in the list. Today, that will mean that if a function and its alias are imported, the alias will be instantiated as a clone of the original. That's likely rare; anyway, it can be addressed subsequently (i.e. have a mode that, instead of cloning, replaces callsites)

File: llvm/test/ThinLTO/X86/Inputs/workload2.ll (hunk 4 above)
Comment (mtrofin) [reply llvm/test/ThinLTO/X86/Inputs/workload2.ll:14]: made it external.

File: llvm/test/ThinLTO/X86/workload.ll (hunk 5 above)
Comment (mtrofin) [reply llvm/test/ThinLTO/X86/workload.ll:53]: `split-file` should take care of that - ptal

File: llvm/lib/Transforms/IPO/FunctionImport.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:517]: simplified the logic here, ptal

File: llvm/lib/Transforms/IPO/FunctionImport.cpp (hunk 12 above)
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp]: (summary of offline discussion) we do want to do the the same for interposable as for non-interposable, so simplified the logic. Ptal.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp]: done.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp [hunk 13]
```diff
@@ -369,29 +373,238 @@ class GlobalsImporter final {
...
+          LLVM_DEBUG(dbgs() << "[Workload] " << Callee << " not found\n");
+          continue;
+        }
+        Set.insert(ElemIt->second);
+      }
+      LLVM_DEBUG(dbgs() << "[Workload] Root: " << Root << " we have "
+                        << Set.size() << " distinct callees.\n");
+      LLVM_DEBUG( //
+          for (const auto &VI
+               : Set) {
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:610]: Oh! I thought you meant combining with the LLVM_DEBUG. Auto-formatting insists on it this way, trying to fix...

File: llvm/lib/Transforms/IPO/FunctionImport.cpp [hunk 14]
```diff
@@ -369,29 +387,273 @@ class GlobalsImporter final {
...
+      DenseMap<StringRef, FunctionImporter::ExportSetTy> *ExportLists)
+      : ModuleImportsManager(IsPrevailing, Index, ExportLists) {
+    // Since the workload def uses names, we need a quick lookup
+    // name->ValueInfo.
+    StringMap<ValueInfo> NameToValueInfo;
+    StringSet<> AmbiguousNames;
+    for (auto &I : Index) {
+      ValueInfo VI = Index.getValueInfo(I);
+      if (!NameToValueInfo.insert(std::make_pair(VI.name(), VI)).second)
+        AmbiguousNames.insert(VI.name());
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp]: done

File: llvm/lib/Transforms/IPO/FunctionImport.cpp (hunk 11 above)
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp]: noted that in the comment

File: llvm/test/ThinLTO/X86/workload.ll [hunk 15]
```diff
@@ -1,67 +1,153 @@
+; Set up
```
Comment (teresajohnson) [inline llvm/test/ThinLTO/X86/workload.ll:1]: Nit: add brief comment at the top about what this is testing (e.g. the workload based importing via -thinlto-workload-def.

Also suggest moving the current m1.ll, m2.ll, and m3.ll code to the end so that all the test RUN lines and checks are at the top.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp [hunk 16]
```diff
@@ -451,76 +451,79 @@ class WorkloadImportsManager : public ModuleImportsManager {
...
+        if (!llvm::hasSingleElement(PotentialCandidates))
+          LLVM_DEBUG(
+              dbgs()
+              << "[Workload] Found multiple non-prevailing candidates for "
+              << VI.name()
+              << ". This is unexpected. Are module paths passed to the "
+                 "compiler unique for the modules passed to the linker?");
+        GVS = *PotentialCandidates.begin();
+      } else {
+        assert(llvm::hasSingleElement(PrevailingCandidates));
```
Comment (teresajohnson) [inline llvm/lib/Transforms/IPO/FunctionImport.cpp:513]: This assert seems wrong - we get to the else if PrevailingCandidates is empty. Is this meant to check PotentialCandidates?

File: llvm/lib/Transforms/IPO/FunctionImport.cpp [hunk 17]
```diff
@@ -588,7 +568,11 @@ class WorkloadImportsManager : public ModuleImportsManager {
     std::map<std::string, std::vector<std::string>> WorkloadDefs;
     json::Path::Root NullRoot;
     // The JSON is supposed to contain a dictionary matching the type of
-    // WorkloadDefs.
+    // WorkloadDefs. For example:
```
Comment (teresajohnson) [inline llvm/lib/Transforms/IPO/FunctionImport.cpp:571]: Suggest moving this example to the option definition.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp [hunk 18]
```diff
@@ -451,76 +451,79 @@ class WorkloadImportsManager : public ModuleImportsManager {
...
+      /// but we could specialize it to the workload in other ways. However,
+      /// the linker would drop it in the favor of the prevailing copy.
+      /// Instead, by importing the prevailing variant (assuming also the use
+      /// of `-avail-extern-to-local`), we keep the specialization. We could
+      /// alteranatively make the non-prevailing variant local, but the
+      /// prevailing one is also the one for which we would have previously
+      /// collected profiles, making it preferrable.
+      auto PrevailingCandidates = llvm::make_filter_range(
+          PotentialCandidates, [&](const auto *Candidate) {
+            return IsPrevailing(VI.getGUID(), Candidate);
```
Comment (teresajohnson) [inline llvm/lib/Transforms/IPO/FunctionImport.cpp:501]: There can only be one, so rather than check every single candidate, just quit when the prevailing copy found. But I have a suggestion above for combining this with potential candidates.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp [hunk 19]
```diff
@@ -451,76 +451,79 @@ class WorkloadImportsManager : public ModuleImportsManager {
...
-            LLVM_DEBUG(dbgs() << "[Workload] Skipping " << VI.name() << " from "
-                              << Candidate.second->modulePath()
-                              << " with linkage " << Candidate.second->linkage()
-                              << " because it's not prevailing\n");
-          }
-        }
-      }
-      if (!GVS) {
+      auto PotentialCandidates = llvm::map_range(
+          llvm::make_filter_range(
```
Comment (teresajohnson) [inline llvm/lib/Transforms/IPO/FunctionImport.cpp:468]: I'm a little concerned about the efficiency of having the two make_filter_range invocations, in a case where we may have thousands of copies of a linkonce_odr, for example. Why not just have a single loop and exit as soon as you find the prevailing copy, then check outside the loop to make sure it is importable? You can also keep track of the first valid candidate too in case there isn't a prevailing copy (e.g. a local).

Oh, also, looking at qualifyCalleeCandidates, we will not import anything with interposable linkage (since it isn't inlineable). I think you may want to import in that case given your use case though (failure reason is FunctionImporter::ImportFailureReason::InterposableLinkage)?

File: llvm/lib/Transforms/IPO/FunctionImport.cpp (hunk 17 above)
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:577]: Done

File: llvm/test/ThinLTO/X86/workload.ll (hunk 15 above)
Comment (mtrofin) [reply llvm/test/ThinLTO/X86/workload.ll:3]: Done.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp (hunk 19 above)
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:474]: There's a problem with interposable importing, actually - see `FunctionImportGlobalProcessing::getLinkage`, there's a an assert on `doImportAsDefinition`.

File: llvm/test/ThinLTO/X86/workload.ll [hunk 20]
```diff
@@ -1,67 +1,153 @@
...
+; FIRST-NEXT:   call void @m1_variant
+;
+; FIRST-LABEL:  @m2_f1
+;
+; @noninterposable_f is prevailing in m2 so it will be imported from there. 
+; FIRST-LABEL:  define available_externally void @noninterposable_f
+; FIRST-NEXT:   call void @m2_variant
+;
+; FIRST-LABEL:  define available_externally void @m2_f1_alias
 ; SECOND-LABEL: @m2_f1
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply llvm/test/ThinLTO/X86/workload.ll:88]: Done.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp (hunk 19 above)
Comment (teresajohnson) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:474]: ack, thanks for the explanation.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp (hunk 18 above)
Comment (teresajohnson) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:507]: ack

File: llvm/lib/Transforms/IPO/FunctionImport.cpp (hunk 16 above)
Comment (teresajohnson) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:519]: Oh somehow I misread the earlier if check. Nevermind this comment, but I have another comment then above.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp [hunk 21]
```diff
@@ -369,29 +393,257 @@ class GlobalsImporter final {
...
+      /// of `-avail-extern-to-local`), we keep the specialization. We could
+      /// alteranatively make the non-prevailing variant local, but the
+      /// prevailing one is also the one for which we would have previously
+      /// collected profiles, making it preferrable.
+      auto PrevailingCandidates = llvm::make_filter_range(
+          PotentialCandidates, [&](const auto *Candidate) {
+            return IsPrevailing(VI.getGUID(), Candidate);
+          });
+      if (PrevailingCandidates.empty()) {
+        if (!llvm::hasSingleElement(PotentialCandidates))
```
Comment (teresajohnson) [inline llvm/lib/Transforms/IPO/FunctionImport.cpp:510]: We could in theory have multiple (weak) copies of a symbol when there is no prevailing candidate, if say the prevailing copy was in a native object being linked in. However, we should in theory be marking all of these non-prevailing IR copies dead in that case, in which case they won't be candidates, so I don't *think* we can get here. Maybe verify that the non-prevailing copy that we select below has local linkage?

File: llvm/lib/Transforms/IPO/FunctionImport.cpp (hunk 19 above)
Comment (mtrofin) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:474]: I thought more about interposables. I think they could be imported if also made internal, wdyt? but separate patch.

Comment (teresajohnson) [reply llvm/lib/Transforms/IPO/FunctionImport.cpp:474]: I don't think this would always be safe, because you are making it non-interposable then.

File: llvm/lib/Transforms/IPO/FunctionImport.cpp [hunk 22]
```diff
@@ -369,29 +393,263 @@ class GlobalsImporter final {
...
```
Comment (teresajohnson) [inline llvm/lib/Transforms/IPO/FunctionImport.cpp:510]: Make this only give the error if GVS->hasLocalLinkage() too.

**Review by teresajohnson — APPROVED** [review]
